### ISSUE TITLE
Mask needs to be used while computing padded_label_sum

### ISSUE THREAD

**Issue by ganeshparameswaran** [tensorflow/ranking#278]
https://github.com/tensorflow/ranking/blob/77f280075e220946d0935b44b64770f4dabb37d0/tensorflow_ranking/python/losses_impl.py#L1007

Currently the code is not using mask at all. While computing padded_label_sum, the code needs to make sure that it doesn't include the values of invalid labels. Currently, it's including the epsilon for invalid labels.

Additionally, "padded" is a misnomer. It's really smoothed_label_sum.

**Comment by rjagerman** [comment]
Thanks for reporting the issue. The mask is actually being used as part of the precompute method in the softmax loss:

https://github.com/tensorflow/ranking/blob/77f280075e220946d0935b44b64770f4dabb37d0/tensorflow_ranking/python/losses_impl.py#L991-L992

The fact that mask is being used in `precompute` and not in `_compute_unreduced_loss_impl` is an internal implementation detail that should not matter for the external use of the public `compute`/`compute_per_list`/`compute_unreduced_loss` methods, since all of those should handle the mask correctly.

Furthermore we have a test that validates that the mask is handled correctly for the softmax loss:

https://github.com/tensorflow/ranking/blob/77f280075e220946d0935b44b64770f4dabb37d0/tensorflow_ranking/python/losses_impl_test.py#L1105-L1116

So, from what I can tell there is no problem with the mask and will mark this issue as solved. If you have a reproducible example, preferably in the form of a unit test, that shows the code is not using the mask as expected we can re-open the issue.

**Comment by ganeshparameswaran** [comment]
@rjagerman You don't have a test case where there are no positive labels. That's where the code breaks.

Consider the input labels to `precompute` `[[0, -1], [0, 0]]`, and `mask` `None`

Output labels of precompute will be `[[0, 0], [0, 0]]`, because -1 is an invalid label (used for padding).

Value of labels for softmax after smoothing will be `[[0.5, 0.5], [0.5, 0.5]]`

That's not the desired target top-1 probability.

The top-1 probability for the input `[0, -1]` should be `[1, 0]` because the second element is padded/invalid.

Currently the code doesn't distinguish between the top-1 probabilities of `[0, -1]` and `[0, 0]`. The logits are fine, because the mask is applied in `precompute`.

**Comment by rjagerman** [comment]
Thanks for the clarification @ganeshparameswaran! I can confirm that I am now able to reproduce this problem and will work on a fix.

**Comment by ganeshparameswaran** [comment]
Thaanks @rjagerman!

**Comment by rjagerman** [comment]
Should be fixed in v0.4.1 with commit d553ef3a7cf0029978d546f96b8d5737f4d69030.

**Comment by ganeshparameswaran** [comment]
Thanks again!
